## Preprocessing

In [1]:
%matplotlib inline
# Import our dependencies
import matplotlib as plt
import os

# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import sweetviz as sv

#  Import and read the charity_data.csv.
import pandas as pd 
df = pd.read_csv("../Resources/charity_data.csv")
df.head()

C:\Users\Hamilton\anaconda3\envs\deeplearning\lib\site-packages\tqdm\auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
df = df.drop("EIN", axis=1)
df = df.drop("NAME", axis=1)
df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [3]:
# Determine the number of unique values in each column.
df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# # importing sweetviz
# import sweetviz as sv

# #analyzing the dataset
# sweetviz_report = sv.analyze(df)
# #display the report
# sweetviz_report.show_html('sweetviz_report.html')

In [5]:
# Look at APPLICATION_TYPE value counts for binning
df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
application_types_to_replace = ["T9","T13","T12","T2","T25","T14","T29","T15","T17"]

# Replace in dataframe
for app in application_types_to_replace:
    df['APPLICATION_TYPE'] = df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Look at CLASSIFICATION value counts for binning
class_df = df['CLASSIFICATION'].value_counts()
class_df

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
# class_df[(class_df > 1)]

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = class_df[(class_df < 1000)].index

# Replace in dataframe
for cls in classifications_to_replace:
    df['CLASSIFICATION'] = df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [10]:
df.shape

(34299, 10)

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies_df = pd.get_dummies(df, drop_first = True)

In [12]:
dummies_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  \
0       1     5000              1                     1                     0   
1       1   108590              1                     0                     0   
2       1     5000              0                     0                     0   
3       1     6692              1                     0                     0   
4       1   142590              1                     0                     0   

   APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  \
0                    0                    0                    0   
1                    1                    0                    0   
2                    0                    0                    1   
3                    1                    0                    0   
4                    1                    0                    0   

   APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  ORGANIZATION_Trust  \
0                    0                    0  ...                   0   
1                    0                    0  ...                   0   
2                    0                    0  ...                   0   
3                    0                    0  ...                   1   
4                    0                    0  ...                   1   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_Y  
0                0                  0                         0  
1                0                  0                         0  
2                0                  0                         0  
3                0                  0                         0  
4                0                  0                         0  

[5 rows x 37 columns]

In [13]:
# Split our preprocessed data into our features and target arrays
X = dummies_df.drop("IS_SUCCESSFUL", axis=1)
y = dummies_df["IS_SUCCESSFUL"]

In [14]:
# Split the preprocessed data into a training and testing dataset
# Use sklearn to split dataset
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [15]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [16]:
# pd.DataFrame(X_train_scaled)

## Compile, Train and Evaluate the Model

In [22]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

nn_model = tf.keras.models.Sequential()

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units=18, activation="relu", input_dim=36))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units=9, activation="relu"))

# Another hidden layer
nn_model.add(tf.keras.layers.Dense(units=9, activation="relu"))

# Output layer
nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 18)                666       
                                                                 
 dense_5 (Dense)             (None, 9)                 171       
                                                                 
 dense_6 (Dense)             (None, 9)                 90        
                                                                 
 dense_7 (Dense)             (None, 1)                 10        
                                                                 
Total params: 937
Trainable params: 937
Non-trainable params: 0
_________________________________________________________________


In [23]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [24]:
from keras.callbacks import Callback
mc = tf.keras.callbacks.ModelCheckpoint('weights{epoch:08d}.h5', save_weights_only=True, period=5)

In [25]:
# Train the model
# fit_model = nn_model.fit(X_train_scaled, y_train, epochs=20)
fit_model = nn_model.fit(X_train_scaled, y_train, callbacks=[mc], epochs=20)

Epoch 1/20
804/804 [==============================] - 3s 2ms/step - loss: 0.6161 - accuracy: 0.6862
Epoch 2/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5650 - accuracy: 0.7263
Epoch 3/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5573 - accuracy: 0.7301
Epoch 4/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5541 - accuracy: 0.7319
Epoch 5/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5532 - accuracy: 0.7321
Epoch 6/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5519 - accuracy: 0.7333
Epoch 7/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5513 - accuracy: 0.7327
Epoch 8/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5504 - accuracy: 0.7328
Epoch 9/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5499 - accuracy: 0.7334
Epoch 10/20
804/804 [==============================] - 2s 2ms/step - loss: 0.5493 - accuracy: 0.7336

In [26]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5530 - accuracy: 0.7279 - 701ms/epoch - 3ms/step
Loss: 0.5529696941375732, Accuracy: 0.7279300093650818


In [27]:
# Export our model to HDF5 file
nn_model.save("AlphabetSoupCharity.h6")

INFO:tensorflow:Assets written to: AlphabetSoupCharity.h6\assets


## Sequential model with hyperparameter

In [35]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['tanh', 'relu'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=26,
        max_value=36,
        step=5), activation=activation, input_dim=36))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 3)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=18,
            max_value=36,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [36]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2,
directory=os.path.normpath('C:/')
)

INFO:tensorflow:Reloading Oracle from existing project C:\untitled_project\oracle.json
INFO:tensorflow:Reloading Tuner from C:\untitled_project\tuner0.json


In [30]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

INFO:tensorflow:Oracle triggered exit


In [31]:
# Get best model hyperparameters
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'relu',
 'first_units': 31,
 'num_layers': 2,
 'units_0': 18,
 'units_1': 23,
 'units_2': 28,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 7,
 'tuner/bracket': 2,
 'tuner/round': 2,
 'tuner/trial_id': '4977a298e9142fc5390beb157db90e0a'}

In [32]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5536 - accuracy: 0.7289 - 724ms/epoch - 3ms/step
Loss: 0.5535896420478821, Accuracy: 0.728863000869751


In [21]:
# Get top 3 model hyperparameters and print the values
top_hyper = tuner.get_best_hyperparameters(3)
for param in top_hyper:
    print(param.values)

{'activation': 'relu', 'first_units': 31, 'num_layers': 2, 'units_0': 18, 'units_1': 23, 'units_2': 28, 'tuner/epochs': 20, 'tuner/initial_epoch': 7, 'tuner/bracket': 2, 'tuner/round': 2, 'tuner/trial_id': '4977a298e9142fc5390beb157db90e0a'}
{'activation': 'relu', 'first_units': 31, 'num_layers': 2, 'units_0': 33, 'units_1': 23, 'units_2': 23, 'tuner/epochs': 20, 'tuner/initial_epoch': 0, 'tuner/bracket': 0, 'tuner/round': 0}
{'activation': 'relu', 'first_units': 36, 'num_layers': 1, 'units_0': 33, 'units_1': 33, 'units_2': 28, 'tuner/epochs': 20, 'tuner/initial_epoch': 0, 'tuner/bracket': 0, 'tuner/round': 0}


## Step 4: Write a Report on the Neural Network Model
#### For this part of the Challenge, you’ll write a report on the performance of the deep learning model you created for AlphabetSoup.

#### The report should contain the following:

#### Overview of the analysis: Explain the purpose of this analysis.
##### Using categorical and continuous data features test data modeling with machine learning and nuearl networs to determine if a model can accurate determine success of organization applying for funding from Alphabet Soup.

#### Results: Using bulleted lists and images to support your answers, address the following questions.

### Data Preprocessing

#### What variable(s) are considered the target(s) for your model?
##### * Target for the model was IS_SUCCESSFUL: Was the money used effectively

#### What variable(s) are considered to be the features for your model?
##### The following variables were used as features for our model as they provides a number relavant data points.
##### * AFFILIATION—Affiliated sector of industry
##### * USE_CASE—Use case for funding
##### * ORGANIZATION—Organization type
##### * STATUS—Active status
##### * INCOME_AMT—Income classification
##### * SPECIAL_CONSIDERATIONS—Special consideration for application
##### * ASK_AMT—Funding amount requested
##### In addition to above variables, we also binned below variables to group low volume classification data into a relevant grouping to reduce noise in our model
##### * APPLICATION_TYPE—Alphabet Soup application type
##### * CLASSIFICATION—Government organization classification


#### What variable(s) are neither targets nor features, and should be removed from the input data?
* EIN — Identification columns
* NAME — Identification columns

### Compiling, Training, and Evaluating the Model

#### How many neurons, layers, and activation functions did you select for your neural network model, and why?
##### Once variables were selected and data transformatoin complete there were 36 input dimensions.  To start I tested half the number of inputs in first hidden layer and added two additional hidden layers with 9 neurons each.   This was an efficient starting point to determine accuracy.  Results for inital nearal network model were under the goal of .75
##### Keras models Sequential
###### * 18 neurons first hidden layer
###### * 9 neurons second hidden layer
###### * 9 neurons third hidden layer
##### Results
##### * 268/268 - 1s - loss: 0.5530 - accuracy: 0.7256 - 1s/epoch - 5ms/step
##### * Loss: 0.5529912710189819, Accuracy: 0.7255976796150208In addition we

#### Were you able to achieve the target model performance?
##### No .75 target performance was not achieved

#### What steps did you take to try and increase model performance?
##### In order to test differnt number of hidden layers and neurons and activation functions (tahn, relu), I ran sequential model with hyperparameters.  Best model performance from these steps were the same as initial nearal network model.
##### Results (Relu)
##### * 268/268 - 2s - loss: 0.5536 - accuracy: 0.7289 - 2s/epoch - 6ms/step
##### * Loss: 0.5535896420478821, Accuracy: 0.728863000869751

#### Summary: Summarize the overall results of the deep learning model. Include a recommendation for how a different model could solve this classification problem, and explain your recommendation.
##### Overall neural networks testing done on selected features could not achieve desired target accuracy of 0.75.  Although both neural network model tests were close at 0.727.  I believe additional transformation and binning of data would increase model accuracy and would receommend analyzing features for binning.